### PR TITLE
[Merged by Bors] - feat(CategoryTheory): preservation of pullback/pushout squares

### PR DESCRIPTION
**Pull request by joelriou**
The main result in this PR is that a commutative square `sq : Square C` is a pullback square iff it is mapped to a pullback square in the category of types when we apply the functor `coyoneda.obj X` for any `X : Cᵒᵖ`. In order to do this, it is shown that the proprety `Square.IsPullback` is preserved when a functor preserving this pullback is applied. A similar result is obtained for pushout squares, which required developing the "opposites" API for squares. 

---

- [x] depends on: #15157

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Limits/Preserves/Shapes/Square.lean [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  Square.IsPullback.mk _ (isLimitPullbackConeMapOfIsLimit F sq.fac h.isLimit)
+
+lemma IsPushout.map (h : sq.IsPushout) (F : C ⥤ D) [PreservesColimit (span sq.f₁₂ sq.f₁₃) F] :
+    (sq.map F).IsPushout :=
+  Square.IsPushout.mk _ (isColimitPushoutCoconeMapOfIsColimit F sq.fac h.isColimit)
+
+variable (sq)
+
+lemma isPullback_iff_map_coyoneda_isPullback :
+    sq.IsPullback ↔ ∀ (X : Cᵒᵖ), (sq.map (coyoneda.obj X)).IsPullback := by
```
Comment (erdOne) [inline Mathlib/CategoryTheory/Limits/Preserves/Shapes/Square.lean:50]: Doesn't this follow from the fact that `yoneda` preserves & reflects limits and that limits in functor categories can be checked pointwise? Is it possible to specialize these results to `IsPullback` and get this lemma?

Comment (joelriou) [reply Mathlib/CategoryTheory/Limits/Preserves/Shapes/Square.lean]: I have developped the API in #15157.

**Comment by erdOne** [comment]
Thanks!
maintainer merge

**Comment by mattrobball** [comment]
bors merge

**Pull request closed without merging**